src/dosext/misc/emm.c

```diff
@@ -273,33 +273,44 @@ static void set_map_registers(const struct emm_reg *buf, int pages);
 #define Kdebug1(args)		E_Stub args
 #define Kdebug2(args)		E_Stub args
 
+static int ems_init_done = 0;
+
 void
 ems_helper(void) {
   switch (LWORD(ebx)) {
   case 0:
     E_printf("EMS Init called!\n");
     if (!config.ems_size) {
+      LWORD(eax) = 1;	/* report error */
       LWORD(ebx) = EMS_ERROR_DISABLED_IN_CONFIG;
       return;
     }
+    if (ems_init_done) {
+      LWORD(eax) = 1;	/* report error */
+      LWORD(ebx) = EMS_ERROR_ALREADY_INITIALISED;
+      return;
+    }
     if (HI(ax) < DOSEMU_EMS_DRIVER_MIN_VERSION) {
       error("EMS driver version mismatch: got %i, expected %i, disabling.\n"
             "Please update your ems.sys from the latest dosemu package.\n",
         HI(ax), DOSEMU_EMS_DRIVER_VERSION);
       com_printf("\nEMS driver version mismatch, disabling.\n"
             "Please update your ems.sys from the latest dosemu package.\n"
             "\nPress any key!\n");
+      LWORD(eax) = 1;	/* report error */
       LWORD(ebx) = EMS_ERROR_VERSION_MISMATCH;
       set_IF();
       com_biosgetch();
       clear_IF();
       return;
     }
+
     if (HI(ax) < DOSEMU_EMS_DRIVER_VERSION) {
       warn("EMS driver too old, consider updating %i->%i\n",
         HI(ax), DOSEMU_EMS_DRIVER_VERSION);
       com_printf("EMS driver too old, consider updating.\n");
     }
+    ems_init_done = 1;	/* make sure the driver isn't loaded twice */
```

Comment:
> the 'call HookHimem' is returning AL == 1, which I think means XMS not installed.

As the comment says,
`/* internal driver installed - go out with error */`
IIRC this means that internal driver does
not need any assistance from ems.sys, so
it gets unloaded, but the `XMS_HELPER_XMS_INIT`
is being called, which is enough. In case
of an external himem.sys, ems.sys needs
to stay, so in that case it won't return error.

> But what would you use to check in ems.sys

EMS spec is quite clear on what is to
check, namely this:
http://www.delorie.com/djgpp/doc/rbinter/ix/67/40.html

> Assuming we find another EMS driver already installed
> somehow, does that mean the XMS part of this driver shouldn't install either?

I am pretty sure yes.
Because then you don't even need to
see if you found this one or another one -
it doesn't even matter. If something is
pretending to be ems driver, we go out.